### PR TITLE
Chenyan98/188

### PR DESCRIPTION
**Pull request by ChenYan98**

### CODE REVIEW

File: src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs [hunk 1]
```diff
@@ -89,10 +89,24 @@ public ConfigurationManager(string metadataAddress, HttpClient httpClient)
...
+        /// </summary>
+        /// <param name="metadataAddress">the address to obtain configuration.</param>
+        /// <param name="configRetriever">the <see cref="IConfigurationRetriever{T}"/></param>
+        /// <param name="docRetriever">the <see cref="IDocumentRetriever"/> that reaches out to obtain the configuration.</param>
+        public ConfigurationManager(string metadataAddress, IConfigurationRetriever<T> configRetriever, IDocumentRetriever docRetriever)
+        {
+            if (configRetriever == null)
             {
-                throw new NotImplementedException(typeof(T).FullName);
+                if (!typeof(T).Equals(typeof(WsFederationConfiguration)) && (!typeof(T).Equals(typeof(OpenIdConnectConfiguration))))
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs]: I think we can safely remove this code as this check is in GetConfigurationRetriever.

**Comment by polita** [comment]
Tagging @brockallen, since he requested this feature. Brock, please take a look.

**Comment by brockallen** [comment]
Uhh, looks ok I guess. IIRC the request was to allow a different implementation to be plugged in. Is that what this is for?

**Comment by ChenYan98** [comment]
@brockallen Yes, you can plugged in your own implementation.

**Comment by ChenYan98** [comment]
Leave a comment

File: src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs (hunk 1 above)
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs]: I agree and remove it.

File: src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs [hunk 2]
```diff
@@ -89,12 +89,18 @@ public ConfigurationManager(string metadataAddress, HttpClient httpClient)
...
-            }
+        }
 
+        /// <summary>
+        /// Instantiaties a new <see cref="ConfigurationManager{T}"/> that manages automatic and controls refreshing on configuration data.
+        /// </summary>
+        /// <param name="metadataAddress">the address to obtain configuration.</param>
+        /// <param name="configRetriever">the <see cref="IConfigurationRetriever{T}"/></param>
+        /// <param name="docRetriever">the <see cref="IDocumentRetriever"/> that reaches out to obtain the configuration.</param>
+        public ConfigurationManager(string metadataAddress, IConfigurationRetriever<T> configRetriever, IDocumentRetriever docRetriever)
```
Comment (polita) [inline src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs]: I don't like the design of having a public ctor that takes a parameter, and the ctor behaves differently if the parameter is null versus not null. I recommend enforcing that the configRetriever parameter isn't null in the ctor where it's a parameter. In the new ctor that doesn't have a configRetriever parameter, call the 3-param ctor as you are, but like this:
 
`public ConfigurationManager(string metadataAddress, IDocumentRetriever docRetriever)
: this(metadataAddress, GetConfigurationRetriever(), docRetriever)
          {
          }`

Tagging @brentschmaltz for his feedback.

File: src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs (hunk 1 above)
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs]: Removed

File: src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs (hunk 2 above)
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs]: I like your idea and modified.

Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocol.Extensions/Configuration/ConfigurationManager.cs]: +1

**Comment by brentschmaltz** [comment]
Please squash the 4 commits to 1, push -f, then merge.
:shipit:

**Comment by polita** [comment]
:shipit:

**Pull request merged**